packages/flutter_tools/lib/src/build_info.dart

```diff
@@ -562,8 +563,19 @@ enum AndroidArch {
 
 /// The default set of iOS device architectures to build for.
 List<DarwinArch> defaultIOSArchsForEnvironment(
-    EnvironmentType environmentType) {
-  if (environmentType == EnvironmentType.simulator) {
+  EnvironmentType environmentType,
+  Artifacts artifacts,
+) {
+  // Handle single-arch local engines.
+  if (artifacts is LocalEngineArtifacts) {
+    final String localEngineName = artifacts.localEngineName;
+    if (localEngineName.contains('_arm64')) {
+      return <DarwinArch>[ DarwinArch.arm64 ];
+    }
+    if (localEngineName.contains('_sim')) {
+      return <DarwinArch>[ DarwinArch.x86_64 ];
+    }
```

Comment:
I wonder if we do this enough throughout the tool that it might be worth centralising a `getLocalEngineDarwinArchitecture(String localEngineName)` function, ideally with a shorter, less horrible name.